File: app/csharp/AccountingSystem.Tests/Application/Services/FinancialStatementServiceTest.cs
using System.Data;
using AccountingSystem.Application.Services;
using AccountingSystem.Infrastructure.Persistence.Repositories;
using Dapper;
using FluentAssertions;
using Npgsql;
using Xunit;

namespace AccountingSystem.Tests.Application.Services;

/// <summary>
/// 財務諸表生成サービスのテスト
/// FinancialStatementService の全メソッドをテスト
/// </summary>
public class FinancialStatementServiceTest : DatabaseTestBase
{
    static FinancialStatementServiceTest()
    {
        SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
    }

    private async Task SetUpBalanceSheetDataAsync(NpgsqlConnection connection)
    {
        // 貸借対照表用テストデータ
        // 取引要素区分: 1=資産, 2=負債, 3=純資産
        await connection.ExecuteAsync(@"
            INSERT INTO ""勘定科目マスタ"" (
                ""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""BSPL区分"", ""取引要素区分""
            ) VALUES
            ('1110', '普通預金', '資産', 'B', '1'),
            ('1410', '建物', '資産', 'B', '1'),
            ('2110', '買掛金', '負債', 'B', '2'),
            ('2510', '長期借入金', '負債', 'B', '2'),
            ('3110', '資本金', '純資産', 'B', '3')
            ON CONFLICT (""勘定科目コード"") DO NOTHING
        ");

        var asOfDate = new DateOnly(2024, 1, 31);
        await connection.ExecuteAsync(@"
            INSERT INTO ""日次勘定科目残高"" (
                ""起票日"", ""勘定科目コード"", ""補助科目コード"", ""部門コード"",
                ""プロジェクトコード"", ""決算仕訳フラグ"", ""借方金額"", ""貸方金額""
            ) VALUES
            (@AsOfDate, '1110', '', '', '', 0, 8000000, 0),
            (@AsOfDate, '1410', '', '', '', 0, 2000000, 0),
            (@AsOfDate, '2110', '', '', '', 0, 0, 500000),
            (@AsOfDate, '2510', '', '', '', 0, 0, 4500000),
            (@AsOfDate, '3110', '', '', '', 0, 0, 5000000)
        ", new { AsOfDate = asOfDate });
    }

    private async Task SetUpIncomeStatementDataAsync(NpgsqlConnection connection)
    {
        // 損益計算書用テストデータ
        // 取引要素区分: 4=収益, 5=費用
        // 費用区分: 1=売上原価, 2=販管費
        await connection.ExecuteAsync(@"
            INSERT INTO ""勘定科目マスタ"" (
                ""勘定科目コード"", ""勘定科目名"", ""勘定科目種別"", ""BSPL区分"", ""取引要素区分"", ""費用区分""
            ) VALUES
            ('4110', '売上高', '収益', 'P', '4', NULL),
            ('5110', '売上原価', '費用', 'P', '5', '1'),
            ('6110', '給与手当', '費用', 'P', '5', '2'),
            ('6210', '福利厚生費', '費用', 'P', '5', '2')
            ON CONFLICT (""勘定科目コード"") DO NOTHING
        ");

        var entryDate = new DateOnly(2024, 1, 15);
        await connection.ExecuteAsync(@"
            INSERT INTO ""日次勘定科目残高"" (
                ""起票日"", ""勘定科目コード"", ""補助科目コード"", ""部門コード"",
                ""プロジェクトコード"", ""決算仕訳フラグ"", ""借方金額"", ""貸方金額""
            ) VALUES
            (@EntryDate, '4110', '', '', '', 0, 0, 10000000),
            (@EntryDate, '5110', '', '', '', 0, 6000000, 0),
            (@EntryDate, '6110', '', '', '', 0, 1500000, 0),
            (@EntryDate, '6210', '', '', '', 0, 500000, 0)
        ", new { EntryDate = entryDate });
    }

    #region 貸借対照表生成テスト

    [Fact(DisplayName = "GenerateBalanceSheetAsync - 貸借対照表を生成できる")]
    public async Task GenerateBalanceSheetAsync_貸借対照表を生成できる()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpBalanceSheetDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var asOfDate = new DateOnly(2024, 1, 31);

        // When
        var result = await service.GenerateBalanceSheetAsync(asOfDate);

        // Then
        result.Should().NotBeNull();
        result.AsOfDate.Should().Be(asOfDate);
        result.Assets.Should().NotBeEmpty();
        result.Liabilities.Should().NotBeEmpty();
        result.Equity.Should().NotBeEmpty();
    }

    [Fact(DisplayName = "GenerateBalanceSheetAsync - 貸借平均の原則が成立する")]
    public async Task GenerateBalanceSheetAsync_貸借平均の原則が成立する()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpBalanceSheetDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var asOfDate = new DateOnly(2024, 1, 31);

        // When
        var result = await service.GenerateBalanceSheetAsync(asOfDate);

        // Then: 資産 = 負債 + 純資産
        result.TotalAssets.Should().Be(result.TotalLiabilitiesAndEquity);
        result.TotalAssets.Should().Be(10000000m);
        result.TotalLiabilities.Should().Be(5000000m);
        result.TotalEquity.Should().Be(5000000m);
    }

    [Fact(DisplayName = "GenerateBalanceSheetAsync - 構成比率が計算される")]
    public async Task GenerateBalanceSheetAsync_構成比率が計算される()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpBalanceSheetDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var asOfDate = new DateOnly(2024, 1, 31);

        // When
        var result = await service.GenerateBalanceSheetAsync(asOfDate);

        // Then: 普通預金 8,000,000 / 10,000,000 = 80%
        var deposit = result.Assets.First(a => a.AccountCode == "1110");
        deposit.Percentage.Should().Be(80.00m);
    }

    #endregion

    #region 損益計算書生成テスト

    [Fact(DisplayName = "GenerateIncomeStatementAsync - 損益計算書を生成できる")]
    public async Task GenerateIncomeStatementAsync_損益計算書を生成できる()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpIncomeStatementDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var fromDate = new DateOnly(2024, 1, 1);
        var toDate = new DateOnly(2024, 1, 31);

        // When
        var result = await service.GenerateIncomeStatementAsync(fromDate, toDate);

        // Then
        result.Should().NotBeNull();
        result.FromDate.Should().Be(fromDate);
        result.ToDate.Should().Be(toDate);
        result.Revenues.Should().NotBeEmpty();
        result.Expenses.Should().NotBeEmpty();
    }

    [Fact(DisplayName = "GenerateIncomeStatementAsync - 利益項目が正しく計算される")]
    public async Task GenerateIncomeStatementAsync_利益項目が正しく計算される()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpIncomeStatementDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var fromDate = new DateOnly(2024, 1, 1);
        var toDate = new DateOnly(2024, 1, 31);

        // When
        var result = await service.GenerateIncomeStatementAsync(fromDate, toDate);

        // Then
        // 売上高: 10,000,000
        result.TotalRevenues.Should().Be(10000000m);
        // 総費用: 6,000,000 + 1,500,000 + 500,000 = 8,000,000
        result.TotalExpenses.Should().Be(8000000m);
        // 売上総利益: 10,000,000 - 6,000,000 = 4,000,000
        result.GrossProfit.Should().Be(4000000m);
        // 営業利益: 4,000,000 - 2,000,000 = 2,000,000
        result.OperatingIncome.Should().Be(2000000m);
        // 当期純利益: 10,000,000 - 8,000,000 = 2,000,000
        result.NetIncome.Should().Be(2000000m);
    }

    [Fact(DisplayName = "GenerateIncomeStatementAsync - 対売上比率が計算される")]
    public async Task GenerateIncomeStatementAsync_対売上比率が計算される()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpIncomeStatementDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var fromDate = new DateOnly(2024, 1, 1);
        var toDate = new DateOnly(2024, 1, 31);

        // When
        var result = await service.GenerateIncomeStatementAsync(fromDate, toDate);

        // Then: 売上原価 6,000,000 / 10,000,000 = 60%
        var costOfSales = result.Expenses.First(e => e.AccountCode == "5110");
        costOfSales.Percentage.Should().Be(60.00m);
    }

    #endregion

    #region 財務指標計算テスト

    [Fact(DisplayName = "CalculateFinancialRatios - 財務指標を計算できる")]
    public async Task CalculateFinancialRatios_財務指標を計算できる()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpBalanceSheetDataAsync(connection);
        await SetUpIncomeStatementDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var asOfDate = new DateOnly(2024, 1, 31);
        var fromDate = new DateOnly(2024, 1, 1);

        var balanceSheet = await service.GenerateBalanceSheetAsync(asOfDate);
        var incomeStatement = await service.GenerateIncomeStatementAsync(fromDate, asOfDate);

        // When
        var result = service.CalculateFinancialRatios(balanceSheet, incomeStatement);

        // Then
        result.Should().NotBeNull();
        result.CurrentRatio.Should().BeGreaterThan(0m);
        result.EquityRatio.Should().BeGreaterThan(0m);
        result.GrossProfitMargin.Should().BeGreaterThan(0m);
        result.OperatingProfitMargin.Should().BeGreaterThan(0m);
        result.NetProfitMargin.Should().BeGreaterThan(0m);
        result.Roa.Should().BeGreaterThan(0m);
        result.Roe.Should().BeGreaterThan(0m);
    }

    [Fact(DisplayName = "CalculateFinancialRatios - 安全性指標が正しく計算される")]
    public async Task CalculateFinancialRatios_安全性指標が正しく計算される()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpBalanceSheetDataAsync(connection);
        await SetUpIncomeStatementDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var asOfDate = new DateOnly(2024, 1, 31);
        var fromDate = new DateOnly(2024, 1, 1);

        var balanceSheet = await service.GenerateBalanceSheetAsync(asOfDate);
        var incomeStatement = await service.GenerateIncomeStatementAsync(fromDate, asOfDate);

        // When
        var result = service.CalculateFinancialRatios(balanceSheet, incomeStatement);

        // Then
        // 流動比率: 流動資産8,000,000 / 流動負債500,000 = 1600%
        result.CurrentRatio.Should().Be(1600.00m);
        // 自己資本比率: 純資産5,000,000 / 総資産10,000,000 = 50%
        result.EquityRatio.Should().Be(50.00m);
    }

    [Fact(DisplayName = "CalculateFinancialRatios - 収益性指標が正しく計算される")]
    public async Task CalculateFinancialRatios_収益性指標が正しく計算される()
    {
        // Given
        await using var connection = new NpgsqlConnection(ConnectionString);
        await connection.OpenAsync();
        await SetUpBalanceSheetDataAsync(connection);
        await SetUpIncomeStatementDataAsync(connection);

        var service = new FinancialStatementService(ConnectionString);
        var asOfDate = new DateOnly(2024, 1, 31);
        var fromDate = new DateOnly(2024, 1, 1);

        var balanceSheet = await service.GenerateBalanceSheetAsync(asOfDate);
        var incomeStatement = await service.GenerateIncomeStatementAsync(fromDate, asOfDate);

        // When
        var result = service.CalculateFinancialRatios(balanceSheet, incomeStatement);

        // Then
        // 売上総利益率: 4,000,000 / 10,000,000 = 40%
        result.GrossProfitMargin.Should().Be(40.00m);
        // 営業利益率: 2,000,000 / 10,000,000 = 20%
        result.OperatingProfitMargin.Should().Be(20.00m);
        // 当期純利益率: 2,000,000 / 10,000,000 = 20%
        result.NetProfitMargin.Should().Be(20.00m);
        // ROA: 2,000,000 / 10,000,000 = 20%
        result.Roa.Should().Be(20.00m);
        // ROE: 2,000,000 / 5,000,000 = 40%
        result.Roe.Should().Be(40.00m);
    }

    #endregion
}

File: app/fsharp/AccountingSystem.Tests/Application/FinancialStatementServiceTest.fs
module AccountingSystem.Tests.Application.FinancialStatementServiceTest

open System
open System.Threading.Tasks
open Xunit
open FsUnit.Xunit
open AccountingSystem.Domain.Models
open AccountingSystem.Application.Port.In
open AccountingSystem.Application.Port.Out
open AccountingSystem.Application.Services

/// <summary>
/// モックリポジトリ
/// </summary>
type MockFinancialStatementRepository(balanceSheet: BalanceSheet, incomeStatement: IncomeStatement) =
    interface IFinancialStatementRepository with
        member _.GenerateBalanceSheetAsync(_asOfDate: DateTime) =
            Task.FromResult(balanceSheet)

        member _.GenerateIncomeStatementAsync(_fromDate: DateTime, _toDate: DateTime) =
            Task.FromResult(incomeStatement)

/// <summary>
/// FinancialStatementService のテスト
/// </summary>
type FinancialStatementServiceTest() =

    /// テスト用の貸借対照表を作成
    member private _.CreateBalanceSheet() : BalanceSheet =
        {
            AsOfDate = DateTime(2024, 1, 31)
            Assets = [
                { AccountCode = "1110"; AccountName = "普通預金"; Balance = 8000000M; Percentage = 80M }
                { AccountCode = "1410"; AccountName = "建物"; Balance = 2000000M; Percentage = 20M }
            ]
            Liabilities = [
                { AccountCode = "2110"; AccountName = "買掛金"; Balance = 500000M; Percentage = 5M }
                { AccountCode = "2510"; AccountName = "長期借入金"; Balance = 4500000M; Percentage = 45M }
            ]
            Equity = [
                { AccountCode = "3110"; AccountName = "資本金"; Balance = 5000000M; Percentage = 50M }
            ]
            TotalAssets = 10000000M
            TotalLiabilities = 5000000M
            TotalEquity = 5000000M
            TotalLiabilitiesAndEquity = 10000000M
        }

    /// テスト用の損益計算書を作成
    member private _.CreateIncomeStatement() : IncomeStatement =
        {
            FromDate = DateTime(2024, 1, 1)
            ToDate = DateTime(2024, 1, 31)
            Revenues = [
                { AccountCode = "4110"; AccountName = "売上高"; Balance = 10000000M; Percentage = 100M }
            ]
            Expenses = [
                { AccountCode = "5110"; AccountName = "売上原価"; Balance = 6000000M; Percentage = 60M }
                { AccountCode = "6110"; AccountName = "給与手当"; Balance = 2000000M; Percentage = 20M }
                { AccountCode = "6120"; AccountName = "法定福利費"; Balance = 500000M; Percentage = 5M }
            ]
            GrossProfit = 4000000M
            OperatingIncome = 1500000M
            NetIncome = 1500000M
            TotalRevenues = 10000000M
            TotalExpenses = 8500000M
        }

    [<Fact>]
    member this.``貸借対照表を生成できる``() =
        task {
            // Given
            let balanceSheet = this.CreateBalanceSheet()
            let incomeStatement = this.CreateIncomeStatement()
            let repository = MockFinancialStatementRepository(balanceSheet, incomeStatement)
            let service = FinancialStatementService(repository) :> IFinancialStatementUseCase
            let asOfDate = DateTime(2024, 1, 31)

            // When
            let! result = service.GenerateBalanceSheetAsync(asOfDate)

            // Then
            result.AsOfDate |> should equal asOfDate
            result.TotalAssets |> should equal 10000000M
            result.TotalLiabilities |> should equal 5000000M
            result.TotalEquity |> should equal 5000000M
        }

    [<Fact>]
    member this.``損益計算書を生成できる``() =
        task {
            // Given
            let balanceSheet = this.CreateBalanceSheet()
            let incomeStatement = this.CreateIncomeStatement()
            let repository = MockFinancialStatementRepository(balanceSheet, incomeStatement)
            let service = FinancialStatementService(repository) :> IFinancialStatementUseCase
            let fromDate = DateTime(2024, 1, 1)
            let toDate = DateTime(2024, 1, 31)

            // When
            let! result = service.GenerateIncomeStatementAsync(fromDate, toDate)

            // Then
            result.FromDate |> should equal fromDate
            result.ToDate |> should equal toDate
            result.TotalRevenues |> should equal 10000000M
            result.TotalExpenses |> should equal 8500000M
            result.NetIncome |> should equal 1500000M
        }

    [<Fact>]
    member this.``財務指標を計算できる``() =
        task {
            // Given
            let balanceSheet = this.CreateBalanceSheet()
            let incomeStatement = this.CreateIncomeStatement()
            let repository = MockFinancialStatementRepository(balanceSheet, incomeStatement)
            let service = FinancialStatementService(repository) :> IFinancialStatementUseCase
            let asOfDate = DateTime(2024, 1, 31)
            let fromDate = DateTime(2024, 1, 1)
            let toDate = DateTime(2024, 1, 31)

            // When
            let! ratios = service.CalculateFinancialRatiosAsync(asOfDate, fromDate, toDate)

            // Then
            // 流動比率 = 8,000,000 / 500,000 × 100 = 1600%
            ratios.CurrentRatio |> should equal 1600.00M
            // 自己資本比率 = 5,000,000 / 10,000,000 × 100 = 50%
            ratios.EquityRatio |> should equal 50.00M
            // 売上総利益率 = 4,000,000 / 10,000,000 × 100 = 40%
            ratios.GrossProfitMargin |> should equal 40.00M
            // 営業利益率 = 1,500,000 / 10,000,000 × 100 = 15%
            ratios.OperatingProfitMargin |> should equal 15.00M
            // 当期純利益率 = 1,500,000 / 10,000,000 × 100 = 15%
            ratios.NetProfitMargin |> should equal 15.00M
            // ROA = 1,500,000 / 10,000,000 × 100 = 15%
            ratios.Roa |> should equal 15.00M
            // ROE = 1,500,000 / 5,000,000 × 100 = 30%
            ratios.Roe |> should equal 30.00M
        }
